### ISSUE TITLE
Replace ultralytics backend with YOLOX in `caatinga_vision` and validate end-to-end camera pipeline

### ISSUE THREAD

**Issue by samuk** [Agroecology-Lab/feldfreund_devkit_ros#16]
**Labels:** `enhancement` `vision` `caatinga` `good-first-issue`
**Milestone:** Vision Stack v1

---

## Background

`caatinga_vision/yolo_inference_node.py` currently drives inference through the `ultralytics` package (YOLOv8 API). We want to evaluate [[YOLOX (Megvii)](https://github.com/Megvii-BaseDetection/YOLOX)](https://github.com/Megvii-BaseDetection/YOLOX) as an alternative backend — it is anchor-free, ships YOLOX-Nano/Tiny variants that are more tractable on the Avaota A1 ARM cores, and supports ONNX/TensorRT export paths we will need later for real-time field use.

This ticket covers both the backend swap and the first real end-to-end test of the camera pipeline, since neither has been validated on Jazzy yet.

---

### 1. Install / package YOLOX

YOLOX is **not** on PyPI as a standard package. It must be installed from source:

```bash
git clone https://github.com/Megvii-BaseDetection/YOLOX.git
cd YOLOX
pip install -v -e .
```

Add this as a step in the `devkit_launch` Docker build or as a `rosdep` keys extension — do **not** pin `ultralytics` in `requirements.txt` without a fallback. The two packages can coexist but we should pick one as primary.

**Note:** YOLOX uses `.pth` checkpoint files (PyTorch `state_dict`), not the `.pt` Ultralytics format. Existing `agro_v1` weights are Ultralytics-format and will **not** load directly — see the training/conversion note below.

### 2. Rewrite `yolo_inference_node.py`

The current node calls `YOLO(model_path).predict(frame, ...)` (Ultralytics API). YOLOX has a different entry point:

```python
from yolox.exp import get_exp
from yolox.utils import postprocess
# load exp file, build model, load .pth checkpoint, run model(img_tensor), postprocess with NMS
```

The replacement node needs to:

- Accept `model_path` (`.pth`), `exp_file` (path to the YOLOX exp `.py`, e.g. `yolox_nano.py`), `conf_thresh`, `nms_thresh`, `input_size` as ROS 2 parameters
- Handle the YOLOX image pre-processing: resize to `input_size` (default 416 for Nano/Tiny, 640 for S/M/L/X), letterbox pad, normalise to `[0,1]` (YOLOX does NOT use ImageNet mean/std — it just divides by 255)
- Run `model(img_tensor)` → raw output tensor → `postprocess()` for NMS → bounding boxes in pixel coords
- Publish `Detection2DArray` on the same topic as before so `InfestationAnalyticsNode` needs no changes
- Preserve the atomic overlay cache write (`imwrite` to `.tmp` + `os.replace`) — that detail is correct and should be kept

Keep the two-phase import fallback pattern from the existing node — YOLOX has the same `PYTHONNOUSERSITE` problem in ROS 2 Python environments.

### 3. Model to test with

For initial bring-up, use **YOLOX-Nano** with COCO pretrained weights (0.91 M params, runs on CPU):

```
https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_nano.pth
```

The agro_v1 crop dataset will need to be retrained with the YOLOX training pipeline before production use. For the purposes of this ticket, confirming the pipeline runs correctly with COCO weights is sufficient.

### 4. Update `ia_pipeline.launch.py`

Add the new parameters (`exp_file`, `nms_thresh`, `input_size`) to the launch file. Provide a `yolox_nano.py` exp file in `caatinga_vision/config/` so users don't have to supply an absolute path.

### 5. Fix the hardcoded GPS origin in `infestation_analytics_node.py`

While bringing up the pipeline, also fix the pre-existing bug identified in the code review: `LAT_REF` and `LON_REF` are module-level constants set to a location in Ceará, Brazil. Promote them to ROS 2 parameters:

```python
self.declare_parameter('lat_ref', -5.1873)
self.declare_parameter('lon_ref', -39.2936)
```

This is a blocking bug for any UK field use and is trivial to fix while the file is open.

---

## Validation steps

Work through these in order. Each step should be confirmed before moving to the next.

**Step 1 — bare YOLOX inference (no ROS)**

Confirm YOLOX installs and runs in the Docker container:

```bash
# Inside the devkit container
cd /tmp && wget https://github.com/Megvii-BaseDetection/YOLOX/releases/download/0.1.1rc0/yolox_nano.pth
python3 -c "
from yolox.exp import get_exp
import torch, cv2, numpy as np
exp = get_exp(None, 'yolox-nano')
model = exp.get_model().eval()
ckpt = torch.load('yolox_nano.pth', map_location='cpu')
model.load_state_dict(ckpt['model'])
img = np.zeros((416, 416, 3), dtype=np.uint8)
# should not throw
print('YOLOX Nano loaded OK')
"
```

**Step 2 — camera topic is live**

With the robot or sim running, confirm `/caatinga_vision/camera/image_raw` is publishing:

```bash
ros2 topic hz /caatinga_vision/camera/image_raw
# expect ~30 Hz from USB cam or Gazebo camera plugin
ros2 topic echo /caatinga_vision/camera/image_raw --once --field header
```

If the topic is silent, check `camera_source_node.py` — this node has not been smoke-tested on Jazzy yet and the `cv2.VideoCapture` device index may need to be parameterised.

**Step 3 — inference node publishes detections**

```bash
ros2 launch caatinga_vision ia_pipeline.launch.py \
  model_path:=/tmp/yolox_nano.pth \
  exp_file:=$(ros2 pkg prefix caatinga_vision)/share/caatinga_vision/config/yolox_nano.py

ros2 topic echo /caatinga_vision/detections
# expect Detection2DArray messages (may be empty detections if no COCO objects in view — that is fine)
```

**Step 4 — overlay image is written**

```bash
# After the node has been running for a few seconds:
ls -lh /tmp/caatinga_vision_overlay.jpg
# file should exist, timestamp should be recent
```

Open with `eog` or copy off robot to confirm bounding boxes render correctly on the frame.

**Step 5 — analytics node receives detections**

```bash
ros2 topic echo /caatinga_vision/infestation/status
# Expect JSON with session_id, timestamp_utc, deteccoes_total ticking up
```

Point the camera at something COCO-detectable (person, bottle, chair) and confirm `deteccoes_total` increments and `spray_recommendation` flips to `true`.

**Step 6 — GeoJSON output**

Kill the ia_pipeline launch cleanly (Ctrl-C). Confirm the output directory was written:

```bash
ls ~/agro_robot_ws/logs_rastreabilidade/Pacote_Rastreabilidade_*/
# inspecao_ia.csv  heatmap_infestacao.geojson  ia_status_final.json
```

Open `heatmap_infestacao.geojson` and confirm coordinates are not in Brazil (the lat_ref/lon_ref fix from Step 3 above should take effect here).

---

## Acceptance criteria

- [ ] YOLOX-Nano loads and runs inference in the Docker container without errors
- [ ] `/caatinga_vision/detections` publishes `Detection2DArray` messages at camera frame rate
- [ ] Overlay cache file is written atomically and contains visible bounding boxes
- [ ] `InfestationAnalyticsNode` receives detections and increments counters correctly
- [ ] GeoJSON heatmap is written on clean shutdown
- [ ] `lat_ref` / `lon_ref` are ROS 2 parameters, not hardcoded constants
- [ ] The `exp_file` parameter defaults to the bundled `yolox_nano.py` so no absolute paths are required

---

## Out of scope for this ticket

- Retraining the `agro_v1` crop dataset with YOLOX — separate ticket
- TensorRT / ONNX export — separate ticket
- `sowbot_row_follow/crop_row_node.py` implementation — already tracked separately
- Performance benchmarking on the Avaota A1 ARM cores — after training

---

## References

- Current inference node: `caatinga_vision/caatinga_vision/yolo_inference_node.py`
- Analytics node: `caatinga_vision/caatinga_vision/infestation_analytics_node.py`
- Launch file: `caatinga_vision/launch/ia_pipeline.launch.py`
- YOLOX repo: https://github.com/Megvii-BaseDetection/YOLOX
- YOLOX-ROS2 prior art (Foxy, may be useful reference): https://github.com/Ar-Ray-code/YOLOX-ROS
- Code review findings: `critical_evaluation.md` — LAT_REF/LON_REF bug, missing crop_row_node, dual odom subscription

**Comment by samuk** [comment]
https://github.com/Agroecology-Lab/feldfreund_devkit_ros/pull/26
